rfc/002-fault-tolerance/README.md

```diff
@@ -0,0 +1,433 @@
+
+Here are some incomplete notes of taking a first look:
+
+- Multi write transactions are shaped like if-then-else statements, they always
+  have a comparison, a list of writes on success and a list of writes on failure
+
+- Watches are possible on single keys and prefixes in the key-space, which would
+  allow us to model our use-case naturally.
+
+- Etcd comes with a gRPC API, making it easy to embed for us. We're already
+  depending on gRPC transitively via Apache Arrow (cf. Flight).
+
+##### Consul
+
+TBD
+
+##### Native Raft
+
+Instead of depending on etcd, we may consider our own actor-based Raft
+implementation. In this case, Raft would be a frontend for a key-value store
+that supports the desired transactional operations natively.
+
+The shape of a metastore would be dedicated CAF service that VAST can speak
+natively to. Not every VAST node should include a Raft metastore, as a typical
+number of Raft nodes remains in the single digits. That is, multiple VAST nodes
+share a single metastore. We would need at least 3 Raft instances to obtain a
+quorum majority when 1 node fails, 5 Raft nodes to tolerate 2 node failures.
+
+VAST already had a
+[working-albeit-minimal](https://github.com/tenzir/vast/pull/926) implementation
+in the past that we removed, given that we didn't need the functionality at the
+time. There are some [open questions on composability of log and state
+machine](https://groups.google.com/g/raft-dev/c/O80OSWrQieo) that we should
+revisit.
+
+## Alternatives
+
+TBD
+
+## Annex
+
+### A1: Considerations regarding the persistence layer as a shared component
+
+As partitions are identified by UUID we consider the write end mostly
+conflict-free. In case a name clash happens anyways the filesystem abstraction
+can just reject one of the writes, this problem can be solved client side by
+renaming the partition. As all this happens before insertion into the metastore
+it isn't observable for queries.
+
+However, the question of when to actually delete data from persistent storage is
+not trivial to answer. As discussed above we want to introduce mandatory
+synchronization for each store listing. The thus-far favored approach of an
+in-memory copy of the metastore state will lag behind a bit, and the states are
+necessarily skewed across multiple nodes as well.
+
+Ideally, a partition shall be removed from persistent storage when it is not
+reachable from any node any more.
+
+Deletions are executed with a delay after the transaction was ACK'ed. The
+deletion would be triggered from the metastore that submitted the transaction,
+on whichever node it lives.
+
+This approach introduces a new tuning knob, namely the duration of the delay
+before the deletion takes place on disk.
```

Comment:
We discussed rollback and time-travel, but can't identify a use case for them.
As described in your comment, this could be added later if the need arises.